// fallout-utility/dist/scripts/getOperatingSystem.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getOperatingSystem = exports.os = void 0;
const os_1 = __importDefault(require("os"));
var os;
(function (os) {
    os[os["WINDOWS"] = 0] = "WINDOWS";
    os[os["LINUX"] = 1] = "LINUX";
    os[os["MACOS"] = 2] = "MACOS";
    os[os["ANDROID"] = 3] = "ANDROID";
    os[os["OPENBSD"] = 4] = "OPENBSD";
    os[os["FREEBSD"] = 5] = "FREEBSD";
    os[os["OTHER"] = 6] = "OTHER";
})(os = exports.os || (exports.os = {}));
function getOperatingSystem() {
    switch (os_1.default.platform()) {
        case 'win32':
            return os.WINDOWS;
        case 'linux':
            return os.LINUX;
        case 'darwin':
            return os.MACOS;
        case 'android':
            return os.ANDROID;
        case 'openbsd':
            return os.OPENBSD;
        case 'freebsd':
            return os.FREEBSD;
        default:
            return os.OTHER;
    }
}
exports.getOperatingSystem = getOperatingSystem;
