// reciple/bin/reciple/classes/builders/InteractionCommandBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InteractionCommandBuilder = void 0;
const builders_1 = require("@discordjs/builders");
class InteractionCommandBuilder extends builders_1.SlashCommandBuilder {
    constructor() {
        super(...arguments);
        this.builder = 'INTERACTION_COMMAND';
        this.cooldown = 0;
        this.requiredBotPermissions = [];
        this.RequiredUserPermissions = [];
        this.allowExecuteInDM = true;
        this.execute = () => { };
    }
    /**
     * Sets the execute cooldown for this command.
     * - `0` means no cooldown
     */
    setCooldown(cooldown) {
        this.cooldown = cooldown;
        return this;
    }
    /**
     * Set required permissions to execute the command
     */
    setRequiredMemberPermissions(...permissions) {
        this.RequiredUserPermissions = permissions;
        return this;
    }
    /**
     * Function when the command is interupted before execution
     */
    setHalt(halt) {
        this.halt = halt ? halt : undefined;
        return this;
    }
    /**
     * Function when the command is executed
     */
    setExecute(execute) {
        if (!execute || typeof execute !== 'function')
            throw new Error('execute must be a function.');
        this.execute = execute;
        return this;
    }
}
exports.InteractionCommandBuilder = InteractionCommandBuilder;
