// reciple/bin/reciple/classes/builders/MessageCommandOptionBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessageCommandOptionBuilder = void 0;
class MessageCommandOptionBuilder {
    constructor() {
        this.name = '';
        this.description = '';
        this.required = false;
        this.validator = () => true;
    }
    /**
     * Set command option name
     */
    setName(name) {
        if (typeof name !== 'string' || !name.match(/^[\w-]{1,32}$/))
            throw new TypeError('name must be a string and match the regex /^[\\w-]{1,32}$/.');
        this.name = name;
        return this;
    }
    /**
     * Set command option description
     */
    setDescription(description) {
        if (!description || typeof description !== 'string')
            throw new TypeError('description must be a string.');
        this.description = description;
        return this;
    }
    /**
     * Set if this option is required
     */
    setRequired(required) {
        if (typeof required !== 'boolean')
            throw new TypeError('required must be a boolean.');
        this.required = required;
        return this;
    }
    /**
     * Set your custom function to validate given value for this option
     */
    setValidator(validator) {
        if (!validator || typeof validator !== 'function')
            throw new TypeError('validator must be a function.');
        this.validator = validator;
        return this;
    }
}
exports.MessageCommandOptionBuilder = MessageCommandOptionBuilder;
